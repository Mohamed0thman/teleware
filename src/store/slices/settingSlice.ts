import {SerializedError, createAsyncThunk, createSlice} from '@reduxjs/toolkit';

interface InitialState {
  lang: 'en' | 'ar' | undefined;
  isRtl: boolean;
  location: {shortCode: string; placeName: string} | null;
  enableNotifee: boolean;
}

export const initialState: InitialState = {
  lang: undefined,
  isRtl: false,
  location: null,
  enableNotifee: false,
};

const settingSlice = createSlice({
  name: 'setting',
  initialState,
  reducers: {
    changeLang: (state, action) => {
      state.lang = action.payload.lang;
      state.isRtl = action.payload.isRtl;
    },
    setLocation: (state, action) => {
      state.location = action.payload;
    },
    setNotifee: (state, action) => {
      state.enableNotifee = action.payload;
    },
  },
  extraReducers: builder => {},
});
export default settingSlice;
export const {changeLang, setLocation, setNotifee} = settingSlice.actions;
