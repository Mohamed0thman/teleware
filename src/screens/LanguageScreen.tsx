import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  I18nManager,
} from 'react-native';
import React, {useEffect} from 'react';
import {useAppDispatch, useAppSelector} from '../store/configureStore';
import {useTranslation} from 'react-i18next';
import RootScreen from '../component/RootScreen';
import {COLORS, FONTS, ICONS, SCALE} from '../constants';
import {StackScreenProps} from '@react-navigation/stack';
import {RootStackParamList} from '../navigation/RootNavigation';
import {checkAndRequestNotificationPermission} from '../utils/permissions';
import {changeLang, setNotifee} from '../store/slices/settingSlice';
import {Typography} from '../component/UI';

const {rw, rh, SC_Width} = SCALE;
const {EgyptFlag, AmircaFlag} = ICONS;

type Props = StackScreenProps<RootStackParamList, 'Language'> & {};

const LanguageScreen = ({navigation}: Props) => {
  const {t, i18n} = useTranslation();
  const {isRtl} = useAppSelector(state => state.setting);
  const dispatch = useAppDispatch();

  const handleOnSelectLang = (lang: 'ar' | 'en', isRtl: boolean) => {
    i18n.changeLanguage(lang);
    dispatch(changeLang({lang, isRtl}));
    navigation.replace('Location');
  };

  useEffect(() => {
    checkAndRequestNotificationPermission().then(() =>
      dispatch(setNotifee(true)),
    );
  }, []);

  //   render button to select lang
  const langSelector = (lang: 'ar' | 'en', title: string, isRtl: boolean) => {
    return (
      <TouchableOpacity
        testID={`${lang}-btn`}
        style={styles.langButton}
        onPress={() => handleOnSelectLang(lang, isRtl)}>
        {/* icon */}
        {lang === 'ar' ? (
          <EgyptFlag width={rw(50)} height={rh(50)} />
        ) : (
          <AmircaFlag width={rw(50)} height={rh(50)} />
        )}
        <Text style={styles.langTitle}>{title}</Text>
      </TouchableOpacity>
    );
  };

  return (
    <RootScreen containerStyle={styles.rootScreen}>
      <Typography testID="screen-title" isRtl={isRtl} style={styles.title}>
        {t('app:titles.langScreen')}
      </Typography>
      <View
        style={[styles.btns, {flexDirection: isRtl ? 'row-reverse' : 'row'}]}>
        {langSelector('en', 'English', false)}
        {langSelector('ar', 'العربيه', true)}
      </View>
    </RootScreen>
  );
};

export default LanguageScreen;

const styles = StyleSheet.create({
  rootScreen: {},
  title: {
    ...FONTS.h2,
    color: COLORS.black,
    textAlign: 'center',
    marginTop: rh(40),
  },
  btns: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    gap: 50,
  },
  langButton: {
    justifyContent: 'center',
    alignItems: 'center',
    width: rw(80),
    height: rh(120),
    borderColor: COLORS.black,
    borderWidth: 1,
  },
  langTitle: {...FONTS.h3, color: COLORS.black},
});
