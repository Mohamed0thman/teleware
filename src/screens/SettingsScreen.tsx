import {StyleSheet, Text, View, TouchableOpacity} from 'react-native';
import React, {useEffect, useState} from 'react';
import RootScreen from '../component/RootScreen';
import {RootStackParamList} from '../navigation/RootNavigation';
import {StackScreenProps} from '@react-navigation/stack';
import {useAppDispatch, useAppSelector} from '../store/configureStore';
import {CustomButton, LineDivider, RadioButton} from '../component/UI';
import notifee, {AuthorizationStatus} from '@notifee/react-native';
import {COLORS, FONTS, ICONS, SCALE, SIZES} from '../constants';
import {setNotifee} from '../store/slices/settingSlice';

const {BellIcon} = ICONS;
const {rh, rw} = SCALE;

type Props = StackScreenProps<RootStackParamList, 'Settings'> & {};

const SettingsScreen = ({navigation}: Props) => {
  const {isRtl, location, enableNotifee} = useAppSelector(
    state => state.setting,
  );

  const dispatch = useAppDispatch();

  return (
    <RootScreen containerStyle={styles.root}>
      <Text style={{...FONTS.h3, color: COLORS.black}}>Your Location</Text>
      <Text style={{...FONTS.h4, color: COLORS.black, marginTop: rh(10)}}>
        {location?.placeName}
      </Text>
      <CustomButton
        testID="toLocation"
        label="Change Location"
        onPress={() => navigation.navigate('Location')}
        containerStyle={{marginTop: rh(20), alignSelf: 'center'}}
      />
      <LineDivider lineStyle={{marginVertical: rh(20)}} />

      <Text style={{...FONTS.h3, color: COLORS.black}}>Change Setting</Text>

      <RadioButton
        icon={<BellIcon width={rw(24)} height={rh(24)} fill={COLORS.white} />}
        isSelected={enableNotifee}
        onPress={() => dispatch(setNotifee(!enableNotifee))}
      />
    </RootScreen>
  );
};

export default SettingsScreen;

const styles = StyleSheet.create({
  root: {paddingHorizontal: rw(SIZES.radius)},
});
