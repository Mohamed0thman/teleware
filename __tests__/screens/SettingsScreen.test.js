import React from 'react';
import LanguageScreen from '../../src/screens/LanguageScreen';

import {fireEvent, waitFor, screen} from '@testing-library/react-native';
import SettingsScreen from '../../src/screens/SettingsScreen';
import renderer, {act} from 'react-test-renderer';
import settingSlice, {
  changeLang,
  initialState,
  setLocation,
} from '../../src/store/slices/settingSlice';

describe('languge screen', () => {
  it('renders correctly', () => {
    renderWithRedux(<SettingsScreen />);
  });

  it('snapshot', () => {
    expect(renderWithRedux(<SettingsScreen />)).toMatchSnapshot();
  });

  it('should navigate to Location', async () => {
    const mockNavigation = {navigate: jest.fn()};

    const {getByTestId, findByTestId} = renderWithRedux(
      <SettingsScreen navigation={mockNavigation} />,
    );

    fireEvent.press(getByTestId('toLocation'));
    expect(mockNavigation.navigate).toHaveBeenCalledTimes(1);
    expect(mockNavigation.navigate).toHaveBeenCalledWith('Location');
  });

  describe('tests for settingSlice', () => {
    test('initialize slice with initialValue', () => {
      const listSliceInit = settingSlice.getInitialState();
      expect(listSliceInit).toBe(initialState);
    });

    test('test change lang', () => {
      const testData = {
        lang: 'ar',
        isRtl: true,
      };

      const afterReducerOperation = settingSlice.reducer(
        initialState,
        changeLang(testData),
      );

      expect(afterReducerOperation).toStrictEqual({
        ...initialState,
        ...testData,
      });
    });

    test('test change location', () => {
      const testData = {location: {shortCode: 'eg', placeName: 'lolololo'}};
      const afterReducerOperation = settingSlice.reducer(
        initialState,
        setLocation(testData.location),
      );

      expect(afterReducerOperation).toStrictEqual({
        ...initialState,
        ...testData,
      });
    });
  });
});
