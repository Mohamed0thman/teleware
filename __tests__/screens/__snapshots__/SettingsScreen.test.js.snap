// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`languge screen snapshot 1`] = `
<View
  style={
    [
      {
        "paddingBottom": 0,
        "paddingTop": 0,
      },
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
      },
      {
        "paddingHorizontal": 24,
      },
    ]
  }
>
  <Text
    style={
      {
        "color": "#000",
        "fontFamily": "Roboto-Bold",
        "fontSize": 20,
      }
    }
  >
    Your Location
  </Text>
  <Text
    style={
      {
        "color": "#000",
        "fontFamily": "Roboto-Medium",
        "fontSize": 16,
        "marginTop": 16.5,
      }
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#007864",
        "height": 82,
        "justifyContent": "center",
        "marginTop": 33,
        "opacity": 1,
        "width": "80%",
      }
    }
    testID="toLocation"
  >
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontFamily": "Roboto-Bold",
          "fontSize": 20,
        }
      }
    >
      Change Location
    </Text>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "#e1e1e1",
          "height": 2,
          "width": "100%",
        },
        {
          "marginVertical": 33,
        },
      ]
    }
  />
  <Text
    style={
      {
        "color": "#000",
        "fontFamily": "Roboto-Bold",
        "fontSize": 20,
      }
    }
  >
    Change Setting
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 80,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#007864",
          "borderRadius": 20,
          "height": 40,
          "justifyContent": "center",
          "width": 40,
        }
      }
    >
      <SvgMock
        fill="#FFFFFF"
        height={39.5}
        width={48}
      />
    </View>
    <View
      style={
        {
          "flex": 1,
          "marginLeft": 12,
        }
      }
    >
      <Text
        style={
          {
            "color": "#007864",
            "fontFamily": "Roboto-Bold",
            "fontSize": 20,
          }
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 40,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "rgba(151, 151, 151, 1)",
            "borderRadius": 3,
            "height": 5,
            "width": "100%",
          }
        }
      />
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "rgba(151, 151, 151, 1)",
            "borderColor": "rgba(151, 151, 151, 1)",
            "borderRadius": 15,
            "borderWidth": 5,
            "height": 25,
            "left": 0,
            "position": "absolute",
            "width": 25,
          }
        }
      />
    </View>
  </View>
</View>
`;
